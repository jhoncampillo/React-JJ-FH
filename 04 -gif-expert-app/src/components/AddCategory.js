import React, { useState } from "react";
import PropTypes from "prop-types";

//importo el estado
//puedo trae la props setCatecories con solo 'props'
export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState(""); //strig vacio

  //creo la funcion para capturar el evento e.target.value
  //se lo asigono al setInputValue
  //con el evento onChage cambio el contenido del input
  const handleInputChangue = (e) => {
    console.log(e.target.value);
    setInputValue(e.target.value);
  };

  //creo en evento enter
  const handleSubmit = (e) => {
    //quito el refresh del navegador
    e.preventDefault();
    console.log("Submit Echo");

    //avaluoo si son espcios
    if (inputValue.trim().length > 2) {
      //console.log(porsp.setCategories); asi seria con props
      //comono tengo acceso al inputValue c
      //adiciojo el inpul a los cata\s que ya hay
      setCategories((cats) => [...cats, inputValue]);
      setInputValue("");
    }
  };
  return (
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChangue}
      ></input>
    </form>
  );
};
//Defino el props types requiered
AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
