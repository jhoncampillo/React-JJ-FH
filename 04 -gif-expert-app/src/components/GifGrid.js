import { useState } from "react";
import React from "react";
//recive el argumento category que viene de input del AddCategory
export const GifGrid = ({ category }) => {
  //creo el estado
  const [count, setCount] = useState(0);

  const getGifs = async () => {
    const url =
      "https://api.giphy.com/v1/gifs/search?q=Goku&limit=10&api_key=rRKAHmsjH1qgSc0SUfBItcBaut4L7DUQ";
    const respuesta = await fetch(url);
    //console.log(respuesta);
    //desetructuring solo traigo la data data.data
    const { data } = await respuesta.json();

    //la recorro con map
    const gifs = data.map((img) => {
      return {
        id: img.id,
        title: img.title,
        url: img.images.downsized_medium.url,
      };
    });
    console.log(gifs);
  };
  getGifs();
  return (
    <div>
      <h3>{category} </h3>
      <h3>{count}</h3>
      <button onClick={() => setCount(count + 1)}></button>
    </div>
  );
};
