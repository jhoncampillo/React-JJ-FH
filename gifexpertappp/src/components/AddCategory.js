import React from "react";
import { useState } from "react";
import PropTypes from "prop-types";
//Traigo el parametro entre corchetes por el desetructuring
const AddCategory = ({ setCategories }) => {
  //creo el useState Para saber llo que la persona escribe en el input

  const [inputValue, setInputValue] = useState(""); //Seimpre es string Vacio

  //creo la funcion del onChangue
  const handleInputChangue = (e) => {
    setInputValue(e.target.value);

    //Log para pruebas
    console.log("handleInputChangue Llamado");
  };

  //CUANDO PULSA ENTER
  const handleSubmit = (e) => {
    //quita el refresh del navegador
    e.preventDefault();
    console.log("Submit OK- Handle Prueba 3 ", { inputValue });

    //valido espacios en blaco con trim
    if (inputValue.trim().length > 2) {
      //ADICIONO EL VALOR DEL INPUT AL LA <li></li>
      //Pondo de primera la ultima categoria solicitada
      setCategories((cats) => [inputValue, ...cats]);
      //reseteo en setinputValue
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      {/* paramtro para Prueba */}
      <p>{inputValue}</p>

      {/* <h2>Addcategory</h2> */}
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChangue}
        placeholder="Busca tu Personaje"
      />
    </form>
  );
};
//Defino el parametro setCategories como required
AddCategory.propTypes = {
  //defino una funcion como required
  setCategories: PropTypes.func.isRequired,
};

export default AddCategory;
