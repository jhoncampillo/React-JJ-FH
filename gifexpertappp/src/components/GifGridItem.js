import React from "react";
//recibo las propiedades del componente GigGrid = id, title, url
//export const GifGridItem = (props) => {
//importo propytypes para props obligatorias - imtp+tab
import PropTypes from "prop-types";

export const GifGridItem = ({ id, title, url }) => {
  //console.log(props); remplazo las propuiedades por id,title,url
  // console.log({ id, title, url });
  return (
    <div className="card animate__animated animate__backInRight">
      {/* {img.title} 
        descargo las imagenes en el DOM
      */}
      <img src={url} alt={title} />
      <p className="title">{title}</p>
    </div>
  );
};
//Defino proptypes Obligatorios
GifGridItem.prototype = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
