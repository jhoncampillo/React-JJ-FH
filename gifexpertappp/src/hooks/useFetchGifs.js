//CUSTOM HOOK - Forma de estraer logica de un componente y reutlizarla de manera mas sencilla
/// Los nombres de los Custom hook empiezan con "use
//funcionan como si fueran funcional components usan efectos , reducer , contextos
import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
  const [state, setState] = useState({
    //estado inicial arreglo vacio y una loading
    data: [],
    loading: true,
  });
  console.log("arranca");
  //ejecuta codigo de manera condicional
  //Le mando un arreglo de dependencia [],
  //Solo se jecuta la Funcion getGifs() cuando el componente es renderizado por primera vez
  //COMO SE CREO UN COMPONENETE NUEVO QUE MANEJA EL HTTPS
  //SE ADCIONAN AL useEffect otros parametros
  // se adicionael parametro de promesas .then
  // Esto se Puede remplazar por .then(setImages)
  useEffect(() => {
    //esperamos a que se ejeute la tareas
    getGifs(category).then((imgs) => {
      setTimeout(() => {
        //cambiamos el estado
        setState({
          data: imgs,
          loading: false,
        });
      }, 3000);
    });
  }, [category]);

  //   setTimeout(() => {
  //     setstate({
  //       data: [1, 2, 3, 4],
  //       loading: true,
  //     });
  //   }, 100000);

  return state;
};
