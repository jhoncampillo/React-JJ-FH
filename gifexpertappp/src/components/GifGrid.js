import React from "react";
//import { useEffect, useState } from "react";
import { GifGridItem } from "./GifGridItem";
//import { getGifs } from "../helpers/getGifs";
import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from "prop-types";

//Recibo el parametro category de GifexpertApp con Destructuring
const GifGrid = ({ category }) => {
  //creo el use state - pero sin el useEffect se ejecuta la funcion en
  //cada evento - se vuelve infinita.- se dispara cada vez que hay un cambio
  //const [images, setImages] = useState([]);

  //USO EL CUSTOM HOOKS
  //const state = useFetchGifs();
  //desestruturing-  data: images = renombra data con images
  const { loading, data: images } = useFetchGifs(category);

  ///console.log(loading);
  //console.log(data);

  //necesito recibir la categoria y hacer una peticion http

  //creo un afuncion

  //getGifs();
  //Capitalizo
  const Category = category.charAt(0).toUpperCase() + category.slice(1);
  console.log(Category);
  return (
    <>
      <h2>{Category}</h2>
      {/* //uso el Custom hooks */}
      {/* //&& si es verdadero muetras de lo contratrio no muestra nada. */}
      {loading && (
        <p className="animate__animated animate__bounceIn">Loading...</p>
      )}
      <div className="card-grid">
        {/* <h3>{count}</h3> */}

        {/* <button onClick={() => setCount(count + 1)}></button> */}

        {/* //terer en cuenta que el return implicito  va con () 
          Puedo desestructurar este map
          */}
        {/* {images.map((img) => (
          <li key={img.id}>{img.title}</li>
        ))} // Le mando el nuevo componente*/}
        {images.map((img) => (
          //   <li key={id}>{title}</li>
          <GifGridItem
            key={img.id}
            // img={img} replazo esto y mando al componente todas las propiedades de img
            //como independientes id,title,url
            {...img}
          />
        ))}
      </div>
    </>
  );
};

//Defino que el el parametro category es requerido
GifGrid.protoTypes = {
  category: PropTypes.string.isRequired,
};

export default GifGrid;
