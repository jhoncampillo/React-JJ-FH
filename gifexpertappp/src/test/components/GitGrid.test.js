//Pruebas de AddCategory
import React from "react";
import { shallow } from "enzyme";
import GifGrid from "../../components/GifGrid";
//Autocompletado
import "@testing-library/jest-dom";
//Importo para simular datos
import { useFetchGifs } from "../../hooks/useFetchGifs";
import { GifGridItem } from "../../components/GifGridItem";
//uso el jest para la simulacion o fingir
jest.mock("../../hooks/useFetchGifs");

//INICIO DE PRUEBAS
describe("Pruebas al componente <GifGrid/>", () => {
  //simulo el parametro
  const category = "One Punch";

  //Primera Prueba
  test("Componente haga Match con el SnapShot ", () => {
    //simulo el retorno de los datos  que inicialmentee en el componente
    //un arreglo vacio data:[] y y loaging:true
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });

    //creo el snapShot
    const wrapper = shallow(<GifGrid category={category} />);
    //Comparo el wrapper con el snapShot
    expect(wrapper).toMatchSnapshot();
  });

  //prueba cunado el useFectch carga la informacion
  test("Debe mostar items cuando se carga  imagenes usefetchs", () => {
    //Mook Fomjir -falcear como si el componente tuibiera informacion
    //probar como si el customHook regreso la Informacion cuando y la  cargo
    //importo el useFetchsGifs pero n o lo llamo
    //creo el snapShot

    //CREO UN OBJETO CON LOS DATOS DE RESPUESTA
    const gifs = [
      {
        id: "ABC",
        url: "https://localhost/xxxx/",
        title: "Cualquier Titulo para prueba",
      },
      {
        id: "HDGDG",
        url: "https://localhost/",
        title: "Cualquier Titulo2 para prueba",
      },
    ];
    console.log(gifs);
    //reasigno los datos de gifs con los que simulo un respuesta
    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={category} />);

    //que esperaria
    expect(wrapper).toMatchSnapshot();
    //Comentamois este Snapshot para mejorar la prueba

    //si el parrafo no existe es por que el loading esta en false
    //es expect evalua si el parrafo existe
    expect(wrapper.find("p").exists()).toBe(false);

    //evaluo si el componente GigGridItem existe-
    //Miro la cantidad de lementos y los comparo con los elementos del gifs
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);

    //ESTA PRUEB AEVALUA EL SNAPCHOT LOS ELEMENTOS QUE RENDERIZA
  });
});
