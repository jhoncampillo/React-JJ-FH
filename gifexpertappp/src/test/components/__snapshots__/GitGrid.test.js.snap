// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al componente <GifGrid/> Componente haga Match con el SnapShot  1`] = `
<Fragment>
  <h2>
    One Punch
  </h2>
  <p
    className="animate__animated animate__bounceIn"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas al componente <GifGrid/> Debe mostar items cuando se carga  imagenes usefetchs 1`] = `
<Fragment>
  <h2>
    One Punch
  </h2>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier Titulo para prueba"
      url="https://localhost/xxxx/"
    />
    <GifGridItem
      id="HDGDG"
      key="HDGDG"
      title="Cualquier Titulo2 para prueba"
      url="https://localhost/"
    />
  </div>
</Fragment>
`;
